src/Symfony/Component/Messenger/Middleware/LockMiddleware.php

```diff
@@ -0,0 +1,82 @@
+<?php
+
+/*
+ * This file is part of the Symfony package.
+ *
+ * (c) Fabien Potencier <fabien@symfony.com>
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Symfony\Component\Messenger\Middleware;
+
+use Symfony\Component\Lock\Key;
+use Symfony\Component\Lock\LockFactory;
+use Symfony\Component\Messenger\Envelope;
+use Symfony\Component\Messenger\Message\LockableMessageInterface;
+use Symfony\Component\Messenger\Message\TTLAwareLockableMessageInterface;
+use Symfony\Component\Messenger\Stamp\LockStamp;
+use Symfony\Component\Messenger\Stamp\ReceivedStamp;
+
+final class LockMiddleware implements MiddlewareInterface
+{
+    private LockFactory $lockFactory;
+
+    public function __construct(LockFactory $lockFactory)
+    {
+        $this->lockFactory = $lockFactory;
+    }
+
+    public function handle(Envelope $envelope, StackInterface $stack): Envelope
+    {
+        $message = $envelope->getMessage();
+        if (!$message instanceof LockableMessageInterface) {
```

Comment:
i'd prefer putting the key in a stamp

Reply:
WDYT @jderusse ?

it's
```
class MyMessage implements LockableMessageInterface
{
    public function getKey() { ... } 
}
```
vs
```
$this->dispatch(new MyMessage(), new LockStamp(new Key('myKey')));
```

I thought it was better in the Message, because the key will be often related to the data in the message.
For instance
```
class MyMessage implements LockableMessageInterface
{
    public construct __construct(public int projectId) {}
    public function getKey() { return self::class . '_' . $this->projectId } 
}
```

Reply:
the stamp brings infinite runtime flexibility, adding them by default to messages can middleware/service concern :) or wait for SF to allow default stamping thru stamps-as-attributes

Reply:
I'm :+1: with `LockableMessageInterface`.

From userland point of view it's simpler to only return a string

It avoids coupling with symfony/lock (people don't have to know how works the middleware)

Also, if the message is dispatched from multiple places, you have to remember that the envelope should contain a stamp.

Reply:
I tried the implementation with the Stamp only with
https://github.com/symfony/symfony/pull/54141/commits/c2331215c9c89d499c7a7b42b9eed44afb873051

And it has some advantages like less code needed (and 0 interfaces) and maybe more flexibility.

We don't provide a `DelayedMessageInterface` to delay message but a DelayStamp instead.
It could be consistent to not provider an interface but just a stamp.

Maybe another feature should be to provide a `StampedMessageInterface` with a method `getDefaultStamps()`.
Other people can have this behavior by doing
```
class MyMessage
{
    public function toEnvelope(): Envelope
    {
        return new Envelope($this, $defaultStamps);
    }
}
```
and writing `$bus->dispatch((new MyMessage())->toEnvelope())`.

Anyway, it's in a separate commit so I can revert/keep it, like you want.